### DISCUSSION TITLE
Eslint wants `useDialogState` inside `useEffect` dependency array

### DISCUSSION THREAD

**Discussion by kwiat1990** [ariakit/ariakit#2341] [Questions] [answered]
```tsx
 const dialog = useDialogState();
 const location = useLocation();

 React.useEffect(() => {
    dialog.setOpen(false);
  }, [location]);
```

It triggers following eslintwarning: `React Hook React.useEffect has a missing dependency: 'dialog'. Either include it or remove the dependency array.eslint[react-hooks/exhaustive-deps](https://github.com/facebook/react/issues/14920)`.

But as soon as I add `dialog` to the dependency array, dialog will open  and then immediately close because it's state changes and thus triggers re-render. For other code I've seen solutions such a `useCallback` or defining function directly inside `useEffect` but this seems to be a different case. Is there a better approach than to disable eslint?

**Comment by diegohaz** [comment, marked as answer]
You can do the following:

```js
 const dialog = useDialogState();
 const location = useLocation();
 const { setOpen } = dialog;

 React.useEffect(() => {
    setOpen(false);
  }, [location, setOpen]);
```

Also, that's one of the reasons we moved to component stores (https://github.com/ariakit/ariakit/issues/1875)

**Reply by kwiat1990** [reply]
Thanks for the helpful feedback. It occurred to me today that all releases marked as `next` weren’t considered by `npm-check` updates. I then updated the package by hand and change refactored my code to use store. It all seems to work besides framer motion exit animations, which don’t occur. Perhaps it’s already answered in another discussion. 

Edit: @diegohaz, for some reason it seems that anmating modal and menu with Framer Motion (enter and exit animations) work only if `alwaysVisible` prop is set on components. Otherwise the exit animation doesn't get applied/isn't visible. Earlier (unofficial releases) it was not a requirement/an issue.

**Reply by diegohaz** [reply]
@kwiat1990 Could you share the code you were using before?

**Reply by kwiat1990** [reply]
My initial approach came form this [discussion](https://github.com/ariakit/ariakit/discussions/1643) and the final code looked following (for brevity I removed some of attributes and classes):
```tsx
export function Modal({
    children,
    className,
    heading,
    state,
    position = "center",
}: Props) {
    return (
    /** use here explicitly the <Portal> to workaround a bug on mobile while using portal prop on <Dialog>:
     *  when modal is placed inside a sticky parent element, then portal container has 0 width
     */
    <AnimatePresence initial={false}>
	{state.open && (
            <Portal className="fixed inset-0">
	        <Body
		    state={state}
		    heading={heading}
		    className={className}
		    position={position}
	        >
		    {children}
	        </Body>
            </Portal>
	)}
    </AnimatePresence>
    );
}

const Backdrop = React.forwardRef(
    (props: {}, ref: React.Ref<HTMLDivElement>) => {
        const animation = useAnimation(AnimationType.fade);
        return <motion.div key="backdrop" ref={ref} {...animation} {...props} />;
    },
);

function Body({ position, className, state, children, heading }: Props) {
    const animation = useAnimation(AnimationType.scale);
    return (
        <Dialog
            backdrop={Backdrop}
            className={className}
            modal={true}
            portal={false}
            state={state}
        >
            <motion.div {...animation}>
                <Button onClick={state.hide}>
	            <RiCloseFill />
                </Button>
                <div>
	            <DialogHeading>
		        {heading}
	            </DialogHeading>
	            {children}
                </div>
            </motion.div>
        </Dialog>
    );
}
```

The above component was used along with the `"ariakit": "^2.0.0-next.32"`. Nonetheless, I get it working with the new version of Ariakit but needed to change some bits of the code to match the one from the [official example](https://ariakit.org/examples/dialog-framer-motion). The key was to set missing `alwaysVisible` prop, which at the first seemed to be something that controls if the modal content should be always rendered into DOM but hidden with `display: none`. Then I saw the very last sentence in the intelli sense, which states it should be used e.g. along with Framer Motion to enable exit animation.
